File: main/java/src/components/IvyXMLParser.java
package components;

import java.awt.Color;
import java.awt.Dimension;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.image.BufferedImage;
import java.io.File;

import javax.swing.JFrame;
import javax.swing.JPanel;

import org.jdom2.Attribute;
import org.jdom2.Document;
import org.jdom2.Element;
import org.jdom2.input.SAXBuilder;

import components.IvyComponent.Alignment;
import components.IvyComponent.Border;
import components.IvyComponent.DimensionParameters;

import exception.*;

public class IvyXMLParser {
	public static IvyComponent parseXML(String fileName){
		SAXBuilder sxb = new SAXBuilder();
		Document document;
		Element racine;
		IvyComponent comp = null;
		
		try{
			document = sxb.build(new File(fileName));

			racine = document.getRootElement();

			comp = parseXMLComponent(racine);
		}catch(Exception e){
			//TODO
		}

		return comp;
	}

	public static IvyComponent parseXMLComponent(Element e) throws IvyException {
		IvyComponent comp = null;

		if(e.getName().equals("Grid")){
			comp = parseXMLGrid(e);
		}
		else if(e.getName().equals("Button")){
			comp = parseXMLButton(e);
		}
		else if(e.getName().equals("Label")){
			comp = parseXMLLabel(e);
		}
		else if(e.getName().equals("List")){
			comp = parseXMLList(e);
		}

		return comp;
	}

	//--- TEXT COMPONENT

	private static IvyButton parseXMLButton(Element e) throws IvyException {
		IvyButton button = new IvyButton();

		for(Attribute attr: e.getAttributes()){
			if(getComponentAttributes(button, attr)){
				continue;
			}

			if(getTextComponentAttributes(button, attr)){
				continue;
			}
		}

		//TODO
		return button;
	}

	private static IvyLabel parseXMLLabel(Element e) throws IvyException {
		IvyLabel label = new IvyLabel();

		for(Attribute attr: e.getAttributes()){
			if(getComponentAttributes(label, attr)){
				continue;
			}

			if(getTextComponentAttributes(label, attr)){
				continue;
			}
		}

		return label;
	}

	private static IvyGrid parseXMLGrid(Element e) throws IvyException {
		IvyGrid grid = new IvyGrid(1, 1);
		String value;
		IvyComponent compChild;

		for(Attribute attr: e.getAttributes()){
			/* Component Attributes */

			if(getComponentAttributes(grid, attr)){
				continue;
			}

			/* Grid Attributes */

			if(attr.getName().equals("rows")){
				value = attr.getValue();

				try{
					grid.setRows(Integer.parseInt(value));
				}catch(Exception ex){
					throw new InvalidAttributeTypeException();
				}
			}
			else if(attr.getName().equals("cols")){
				value = attr.getValue();

				try{
					grid.setCols(Integer.parseInt(value));
				}catch(Exception ex){
					throw new InvalidAttributeTypeException();
				}				
			}
		}

		for(Element child: e.getChildren()){
			compChild = parseXMLComponent(child);

			grid.setComponentAt(compChild, compChild.getRow(), compChild.getCol());
		}

		return grid;
	}

	private static IvyList parseXMLList(Element e) throws IvyException {
		IvyList list = new IvyList();

		for(Attribute attr: e.getAttributes()){
			if(getComponentAttributes(list, attr)){
				continue;
			}
		}

		return list;
	}

	private static boolean getTextComponentAttributes(IvyTextComponent comp, Attribute attr) throws IvyException {
		String value = attr.getValue();

		if(attr.getName().equals("text")){
			comp.setText(value);
		}
		else if(attr.getName().equals("textAlignment")){
			if(value.isEmpty()){
				throw new EmptyAttributeValueException();
			}
			
			if(value.equals("left")){
				comp.setTextAlignment(Alignment.LEFT);
			}
			else if(value.equals("center")){
				comp.setTextAlignment(Alignment.CENTER);
			}
			else if(value.equals("right")){
				comp.setTextAlignment(Alignment.RIGHT);
			}
			else{
				throw new InvalidAttributeException();
			}
		}
		else if(attr.getName().equals("textColor")){
			if(value.isEmpty()){
				throw new EmptyAttributeValueException();
			}
			
			try{
				comp.setBackgroundColor(Color.decode(value));
			}catch(Exception ex){
				throw new InvalidAttributeException();
			}
		}
		else{
			return false;
		}

		return true;
	}

	private static boolean getComponentAttributes(IvyComponent comp, Attribute attr) throws IvyException {
		String value = attr.getValue();

		if(attr.getName().equals("name")){
			//comp.setId()
		}
		else if(attr.getName().equals("width")){
			value = attr.getValue();

			if(value.isEmpty()){
				throw new EmptyAttributeValueException();
			}

			if(value.equals("*")){
				comp.setWidthType(DimensionParameters.FILL);
			}
			else if(value.equals("auto")){
				comp.setWidthType(DimensionParameters.AUTO);
			}
			else{
				comp.setWidthType(DimensionParameters.FIXED);

				try{
					comp.setWidth(Integer.parseInt(value));
				}catch(Exception ex){
					throw new InvalidAttributeTypeException();
				}
			}
		}
		else if(attr.getName().equals("height")){
			value = attr.getValue();

			if(value.isEmpty()){
				throw new EmptyAttributeValueException();
			}

			if(value.equals("*")){
				comp.setHeightType(DimensionParameters.FILL);
			}
			else if(value.equals("auto")){
				comp.setHeightType(DimensionParameters.AUTO);
			}
			else{
				comp.setHeightType(DimensionParameters.FIXED);

				try{
					comp.setHeight(Integer.parseInt(value));
				}catch(Exception ex){
					throw new InvalidAttributeTypeException();
				}
			}
		}
		else if(attr.getName().equals("margin")){
			value = attr.getValue();

			if(value.isEmpty()){
				throw new EmptyAttributeValueException();
			}

			String[] marginValues = value.split(",");

			for(int i=0; i<marginValues.length; i++){
				marginValues[i] = marginValues[i].trim();
			}

			if(marginValues.length == 1){
				try{
					comp.setMargin(Integer.parseInt(marginValues[0]));
				}catch(Exception ex){
					throw new InvalidAttributeTypeException();
				}
			}
			else if(marginValues.length == 2){
				try{
					comp.setMargin(Integer.parseInt(marginValues[0]), 
							Integer.parseInt(marginValues[1]));
				}catch(Exception ex){
					throw new InvalidAttributeTypeException();
				}
			}
			else if(marginValues.length == 4){
				try{
					comp.setMargin(Integer.parseInt(marginValues[0]), 
							Integer.parseInt(marginValues[1]),
							Integer.parseInt(marginValues[2]),
							Integer.parseInt(marginValues[3]));
				}catch(Exception ex){
					throw new InvalidAttributeTypeException();
				}
			}
			else{
				throw new InvalidAttributeValueException();
			}
		}
		else if(attr.getName().equals("row")){
			try{
				comp.setRow(attr.getIntValue());
			}catch(Exception e){
				throw new InvalidAttributeTypeException();
			}
		}
		else if(attr.getName().equals("col")){
			try{
				comp.setCol(attr.getIntValue());
			}catch(Exception e){
				throw new InvalidAttributeTypeException();
			}			
		}
		else if(attr.getName().equals("rowSpan")){
			try{
				comp.setRowSpan(attr.getIntValue());
			}catch(Exception e){
				throw new InvalidAttributeTypeException();
			}
		}
		else if(attr.getName().equals("colSpan")){
			try{
				comp.setColSpan(attr.getIntValue());
			}catch(Exception e){
				throw new InvalidAttributeTypeException();
			}
		}
		else if(attr.getName().equals("border")){
			value = attr.getValue();

			if(value.isEmpty()){
				throw new EmptyAttributeValueException();
			}

			if(value.equals("none")){
				comp.setBorderType(Border.NONE);
			}
			else if(value.equals("squared")){
				comp.setBorderType(Border.SQUARED);
			}
			else if(value.equals("rounded")){
				comp.setBorderType(Border.ROUNDED);
			}
			else{
				throw new InvalidAttributeValueException();
			}
		}
		else if(attr.getName().equals("borderColor")){
			value = attr.getValue();

			if(value.isEmpty()){
				throw new EmptyAttributeValueException();
			}
		}
		else if(attr.getName().equals("backgroundColor")){
			value = attr.getValue();

			if(value.isEmpty()){
				throw new EmptyAttributeValueException();
			}
			
			try{
				comp.setBackgroundColor(Color.decode(value));
			}catch(Exception ex){
				throw new InvalidAttributeException();
			}
		}
		else if(attr.getName().equals("backgroundImage")){
			value = attr.getValue();

			if(value.isEmpty()){
				throw new EmptyAttributeValueException();
			}
		}
		else if(attr.getName().equals("visibility")){
			value = attr.getValue();

			if(value.isEmpty()){
				throw new EmptyAttributeValueException();
			}

			try{
				comp.setVisible(attr.getBooleanValue());
			}catch(Exception ex){
				throw new InvalidAttributeTypeException();
			}
		}
		else{
			return false;
		}

		return true;
	}

	public static void main(String[] args){
		javax.swing.SwingUtilities.invokeLater(new Runnable(){
			public void run(){			

				final IvyComponent comp = parseXML("test.xml");

				JFrame frame = new JFrame();
				/*JPanel content = new JPanel(){
					private static final long serialVersionUID = 1L;

					public void paintComponent(Graphics g){
						if(comp != null){
							BufferedImage bufImage = new BufferedImage(this.getPreferredSize().width,
									this.getPreferredSize().height, BufferedImage.TYPE_INT_RGB);
							//panel.paint(bufImage.createGraphics());

							//comp.draw(g, 0, 0, this.getWidth(), this.getHeight());
							comp.draw(bufImage.createGraphics(), 0, 0, this.getWidth(), this.getHeight());
							g.drawImage(bufImage, 0, 0, bufImage.getWidth(), bufImage.getHeight(), null);
							
							//comp.getComponent().addNotify();
							comp.getComponent().validate();
							//comp.getComponent().setSize(new Dimension(500, 500));
							comp.getComponent().paint(g);
						}
					}
				};*/

				frame.setSize(800, 480);

				//content.setPreferredSize(new Dimension(800, 480));

				/*frame.setContentPane(content);
				frame.pack();*/
				//content.repaint();
				frame.setContentPane(comp.getComponent());
				frame.setResizable(false);
				frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
				frame.setVisible(true);
			}
		});
	}
}

File: main/Ivy/Components/IvyXMLParser.cs
using System.Xml;
using Ivy.Exceptions;
using System;
using System.Windows.Media;
using System.Collections.Generic;

namespace Ivy.Components
{
    class IvyXMLParser
    {
        private static List<string> names;

        public static IvyComponent parseXML(string fileName)
        {
            names = new List<string>();

            lock (names)
            {
                XmlDocument document = new XmlDocument();
                IvyComponent comp = null;

                try
                {
                    document.Load(fileName);

                    comp = parseXMLComponent(document.DocumentElement);
                }
                catch (Exception e)
                {
                    //TODO
                    System.Windows.MessageBox.Show(e.ToString());
                    System.Windows.MessageBox.Show(e.StackTrace);
                }

                return comp;
            }
        }

        public static IvyComponent parseXMLComponent(XmlNode e)
        {
            IvyComponent comp = null;

            if (e.Name.Equals("Grid"))
            {
                comp = parseXMLGrid(e);
            }
            else if (e.Name.Equals("Button"))
            {
                comp = parseXMLButton(e);
            }
            else if (e.Name.Equals("Label"))
            {
                comp = parseXMLLabel(e);
            }
            else if (e.Name.Equals("List"))
            {
                comp = parseXMLList(e);
            }

            return comp;
        }

        //--- TEXT COMPONENT

        private static IvyButton parseXMLButton(XmlNode e)
        {
            IvyButton button = new IvyButton();

            foreach (XmlAttribute attr in e.Attributes)
            {
                if (getComponentAttributes(button, attr))
                {
                    continue;
                }

                if (getTextComponentAttributes(button, attr))
                {
                    continue;
                }
            }

            //TODO
            return button;
        }

        private static IvyLabel parseXMLLabel(XmlNode e)
        {
            IvyLabel label = new IvyLabel();

            foreach (XmlAttribute attr in e.Attributes)
            {
                if (getComponentAttributes(label, attr))
                {
                    continue;
                }

                if (getTextComponentAttributes(label, attr))
                {
                    continue;
                }
            }

            return label;
        }

        private static IvyGrid parseXMLGrid(XmlNode e)
        {
            IvyGrid grid = new IvyGrid(1, 1);
            string value;
            IvyComponent compChild;


            foreach (XmlAttribute attr in e.Attributes)
            {
                /* Component Attributes */

                if (getComponentAttributes(grid, attr))
                {
                    continue;
                }

                /* Grid Attributes */

                if (attr.Name.Equals("rows"))
                {
                    value = attr.Value;

                    try
                    {
                        grid.setRows(int.Parse(value));
                    }
                    catch (Exception ex)
                    {
                        throw new InvalidAttributeTypeException();
                    }
                }
                else if (attr.Name.Equals("cols"))
                {
                    value = attr.Value;

                    try
                    {
                        grid.setCols(int.Parse(value));
                    }
                    catch (Exception ex)
                    {
                        throw new InvalidAttributeTypeException();
                    }
                }
            }

            if (e.HasChildNodes)
            {
                foreach (XmlNode child in e.ChildNodes)
                {
                    compChild = parseXMLComponent(child);
                    grid.setComponentAt(compChild, compChild.getRow(), compChild.getCol());
                }
            }

            return grid;
        }

        private static IvyList parseXMLList(XmlNode e)
        {
            IvyList list = new IvyList();

            foreach (XmlAttribute attr in e.Attributes)
            {
                if (getComponentAttributes(list, attr))
                {
                    continue;
                }
            }

            return list;
        }

        private static bool getTextComponentAttributes(IvyTextComponent comp, XmlAttribute attr)
        {
            String value = attr.Value;

            if (attr.Name.Equals("text"))
            {
                comp.setText(value);
            }
            else if (attr.Name.Equals("textAlignment"))
            {
                if (value == string.Empty)
                {
                    throw new EmptyAttributeValueException();
                }

                if (value.Equals("left"))
                {
                    comp.setTextAlignment(IvyComponent.Alignment.LEFT);
                }
                else if (value.Equals("center"))
                {
                    comp.setTextAlignment(IvyComponent.Alignment.CENTER);
                }
                else if (value.Equals("right"))
                {
                    comp.setTextAlignment(IvyComponent.Alignment.RIGHT);
                }
                else
                {
                    throw new InvalidAttributeException();
                }
            }
            else if (attr.Name.Equals("verticalTextAlignment"))
            {
                if (value == string.Empty)
                {
                    throw new EmptyAttributeValueException();
                }

                if (value.Equals("top"))
                {
                    comp.setVerticalTextAlignment(IvyComponent.Alignment.TOP);
                }
                else if (value.Equals("center"))
                {
                    comp.setVerticalTextAlignment(IvyComponent.Alignment.CENTER);
                }
                else if (value.Equals("right"))
                {
                    comp.setVerticalTextAlignment(IvyComponent.Alignment.BOTTOM);
                }
                else
                {
                    throw new InvalidAttributeException();
                }
            }
            else if (attr.Name.Equals("textColor"))
            {
                if (value == string.Empty)
                {
                    throw new EmptyAttributeValueException();
                }

                try
                {
                    comp.setTextColor((SolidColorBrush)new BrushConverter().ConvertFromString(value));
                }
                catch (Exception ex)
                {
                    throw new InvalidAttributeException();
                }
            }
            else if (attr.Name.Equals("padding"))
            {
                value = attr.Value;

                if (value == string.Empty)
                {
                    throw new EmptyAttributeValueException();
                }

                String[] paddingValues = value.Split(',');

                for (int i = 0; i < paddingValues.Length; i++)
                {
                    paddingValues[i] = paddingValues[i].Trim();
                }

                if (paddingValues.Length == 1)
                {
                    try
                    {
                        comp.setPadding(int.Parse(paddingValues[0]));
                    }
                    catch (Exception ex)
                    {
                        throw new InvalidAttributeTypeException();
                    }
                }
                else if (paddingValues.Length == 2)
                {
                    try
                    {
                        comp.setPadding(int.Parse(paddingValues[0]),
                                int.Parse(paddingValues[1]));
                    }
                    catch (Exception ex)
                    {
                        throw new InvalidAttributeTypeException();
                    }
                }
                else if (paddingValues.Length == 4)
                {
                    try
                    {
                        comp.setPadding(int.Parse(paddingValues[0]),
                                int.Parse(paddingValues[1]),
                                int.Parse(paddingValues[2]),
                                int.Parse(paddingValues[3]));
                    }
                    catch (Exception ex)
                    {
                        throw new InvalidAttributeTypeException();
                    }
                }
                else
                {
                    throw new InvalidAttributeValueException();
                }
            }
            else
            {
                return false;
            }

            return true;
        }

        private static bool getComponentAttributes(IvyComponent comp, XmlAttribute attr)
        {
            String value = attr.Value;

            if (attr.Name.Equals("name"))
            {
                value = attr.Value;

                if (value == string.Empty)
                {
                    throw new EmptyAttributeValueException();
                }

                if (names.Contains(value))
                {
                    throw new DuplicateNameException();
                }

                comp.setName(value);
            }
            else if (attr.Name.Equals("width"))
            {
                value = attr.Value;

                if (value == string.Empty)
                {
                    throw new EmptyAttributeValueException();
                }

                if (value.Equals("*"))
                {
                    comp.setWidthType(IvyComponent.DimensionParameters.FILL);
                }
                else if (value.Equals("auto"))
                {
                    comp.setWidthType(IvyComponent.DimensionParameters.AUTO);
                }
                else
                {
                    comp.setWidthType(IvyComponent.DimensionParameters.FIXED);

                    try
                    {
                        comp.setWidth(int.Parse(value));
                    }
                    catch (Exception ex)
                    {
                        throw new InvalidAttributeTypeException();
                    }
                }
            }
            else if (attr.Name.Equals("height"))
            {
                value = attr.Value;

                if (value == string.Empty)
                {
                    throw new EmptyAttributeValueException();
                }

                if (value.Equals("*"))
                {
                    comp.setHeightType(IvyComponent.DimensionParameters.FILL);
                }
                else if (value.Equals("auto"))
                {
                    comp.setHeightType(IvyComponent.DimensionParameters.AUTO);
                }
                else
                {
                    comp.setHeightType(IvyComponent.DimensionParameters.FIXED);

                    try
                    {
                        comp.setHeight(int.Parse(value));
                    }
                    catch (Exception ex)
                    {
                        throw new InvalidAttributeTypeException();
                    }
                }
            }
            else if (attr.Name.Equals("margin"))
            {
                value = attr.Value;

                if (value == string.Empty)
                {
                    throw new EmptyAttributeValueException();
                }

                String[] marginValues = value.Split(',');

                for (int i = 0; i < marginValues.Length; i++)
                {
                    marginValues[i] = marginValues[i].Trim();
                }

                if (marginValues.Length == 1)
                {
                    try
                    {
                        comp.setMargin(int.Parse(marginValues[0]));
                    }
                    catch (Exception ex)
                    {
                        throw new InvalidAttributeTypeException();
                    }
                }
                else if (marginValues.Length == 2)
                {
                    try
                    {
                        comp.setMargin(int.Parse(marginValues[0]),
                                int.Parse(marginValues[1]));
                    }
                    catch (Exception ex)
                    {
                        throw new InvalidAttributeTypeException();
                    }
                }
                else if (marginValues.Length == 4)
                {
                    try
                    {
                        comp.setMargin(int.Parse(marginValues[0]),
                                int.Parse(marginValues[1]),
                                int.Parse(marginValues[2]),
                                int.Parse(marginValues[3]));
                    }
                    catch (Exception ex)
                    {
                        throw new InvalidAttributeTypeException();
                    }
                }
                else
                {
                    throw new InvalidAttributeValueException();
                }
            }
            else if (attr.Name.Equals("row"))
            {
                try
                {
                    comp.setRow(int.Parse(attr.Value));
                }
                catch (Exception e)
                {
                    throw new InvalidAttributeTypeException();
                }
            }
            else if (attr.Name.Equals("col"))
            {
                try
                {
                    comp.setCol(int.Parse(attr.Value));
                }
                catch (Exception ex)
                {
                    throw new InvalidAttributeTypeException();
                }
            }
            else if (attr.Name.Equals("rowSpan"))
            {
                try
                {
                    comp.setRowSpan(int.Parse(attr.Value));
                }
                catch (Exception ex)
                {
                    throw new InvalidAttributeTypeException();
                }
            }
            else if (attr.Name.Equals("colSpan"))
            {
                try
                {
                    comp.setColSpan(int.Parse(attr.Value));
                }
                catch (Exception ex)
                {
                    throw new InvalidAttributeTypeException();
                }
            }
            else if (attr.Name.Equals("border"))
            {
                value = attr.Value;

                if (value == string.Empty)
                {
                    throw new EmptyAttributeValueException();
                }

                if (value.Equals("none"))
                {
                    comp.setBorderType(IvyComponent.Border.NONE);
                }
                else if (value.Equals("squared"))
                {
                    comp.setBorderType(IvyComponent.Border.SQUARED);
                }
                else if (value.Equals("rounded"))
                {
                    comp.setBorderType(IvyComponent.Border.ROUNDED);
                }
                else
                {
                    throw new InvalidAttributeValueException();
                }
            }
            else if (attr.Name.Equals("borderColor"))
            {
                value = attr.Value;

                if (value == string.Empty)
                {
                    throw new EmptyAttributeValueException();
                }

                /* TODO */
            }
            else if (attr.Name.Equals("backgroundColor"))
            {
                value = attr.Value;

                if (value == string.Empty)
                {
                    throw new EmptyAttributeValueException();
                }

                try
                {
                    comp.setBackgroundColor((SolidColorBrush)new BrushConverter().ConvertFromString(value));
                }
                catch (Exception ex)
                {
                    throw new InvalidAttributeException();
                }
            }
            else if (attr.Name.Equals("backgroundImage"))
            {
                value = attr.Value;

                if (value == string.Empty)
                {
                    throw new EmptyAttributeValueException();
                }

                try
                {
                    comp.setBackgroundImage(value);
                }
                catch (Exception ex)
                {
                    throw new NoSuchResourceException();
                }
            }
            else if (attr.Name.Equals("visible"))
            {
                value = attr.Value;

                if (value == string.Empty)
                {
                    throw new EmptyAttributeValueException();
                }

                try
                {
                    comp.setVisible(bool.Parse(value));
                }
                catch (Exception ex)
                {
                    throw new InvalidAttributeTypeException();
                }
            }
            else
            {
                return false;
            }

            return true;
        }
    }
}
